realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -422,6 +422,8 @@ private static void initializeSyncedRealm(Realm realm) {
         } finally {
             if (commitChanges) {
                 realm.commitTransaction(false);
+            } else {
+                realm.cancelTransaction();
```

Comment:
I think it's safe to check first if we're in a transaction `realm.isInTransaction()`